File: app/src/main/java/com/chichiangho/base/utils/Logger.java
package com.chichiangho.base.utils;

import android.os.Environment;
import android.util.Log;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;

/**
 * 封装原有的日志信息，通过配置控制不同类型的日志输出。
 */
public final class Logger {

    private static final boolean IS_VERBOSE = true;
    public static final boolean IS_DEBUG = true;
    private static final boolean IS_INFO = true;
    private static final boolean IS_WARN = true;
    private static final boolean IS_ERROR = true;
    private static final boolean ENABLE_LOGTO_FILE = false;
    private static final String PREFIX = "--->";

    public static void v(String tag, String msg) {
        if (IS_VERBOSE) {
            StackTraceElement targetStackTraceElement = getTargetStackTraceElement();
            Log.v(tag, PREFIX + msg + "(" + targetStackTraceElement.getFileName() + ":"
                    + targetStackTraceElement.getLineNumber() + ")");
        }
    }

    public static void debug(String tag, String msg) {
        if (IS_DEBUG) {
            Log.d(tag, PREFIX + msg);
        }
    }

    public static void d(String tag, String msg) {
        if (IS_DEBUG) {
            StackTraceElement targetStackTraceElement = getTargetStackTraceElement();
            Log.d(tag, PREFIX + msg + "(" + targetStackTraceElement.getFileName() + ":"
                    + targetStackTraceElement.getLineNumber() + ")");
        }
    }

    public static void i(String tag, String msg) {
        if (IS_INFO) {
            StackTraceElement targetStackTraceElement = getTargetStackTraceElement();
            Log.i(tag, PREFIX + msg + "(" + targetStackTraceElement.getFileName() + ":"
                    + targetStackTraceElement.getLineNumber() + ")");
        }
    }

    public static void w(String tag, String msg) {
        if (IS_WARN) {
            StackTraceElement targetStackTraceElement = getTargetStackTraceElement();
            Log.w(tag, PREFIX + msg + "(" + targetStackTraceElement.getFileName() + ":"
                    + targetStackTraceElement.getLineNumber() + ")");
        }
    }

    public static void e(String tag, String msg) {
        if (IS_ERROR) {
            StackTraceElement targetStackTraceElement = getTargetStackTraceElement();
            Log.e(tag, PREFIX + msg + "(" + targetStackTraceElement.getFileName() + ":"
                    + targetStackTraceElement.getLineNumber() + ")");
        }
    }

    public static void json(String tag, String msg) {
        if (IS_DEBUG) {
            getJson(tag, msg);
        }
    }

    /**
     * Print log but print 3000 characters at most in every line.
     *
     * @param tag
     * @param msg
     */
    public static void printLog(String tag, String msg) {
        if (IS_DEBUG && msg != null) {
            int count = msg.length();
            int lineCount = 3000;
            for (int i = 0; i < count; i = i + lineCount) {
                if (i + lineCount <= count) {
                    Log.d(tag, msg.substring(i, i + lineCount));
                } else {
                    Log.d(tag, msg.substring(i, count));
                }
            }
        }
    }

    /**
     * Print log but print 3000 characters at most in every line.
     *
     * @param tag
     * @param msg
     */
    public static void printLog2(String tag, String msg) {
        if (IS_DEBUG && msg != null) {
            int count = msg.length();
            int lineCount = 3000;
            for (int i = 0; i < count; i = i + lineCount) {
                if (i + lineCount <= count) {
                    Log.d(tag, msg.substring(i, i + lineCount));
                } else {
                    Log.d(tag, msg.substring(i, count));
                }
            }
        }
    }

    public static void logToFile(String time, Object msg, String fileName) {
        if (null == msg) {
            return;
        }
        if (ENABLE_LOGTO_FILE) {
            writeToSDCard(fileName, "[" + time + "] " + msg.toString() + "\n");
        }
    }

    public static void cleanLogFile(String fileName) {
        if (Environment.getExternalStorageState().equals(Environment.MEDIA_MOUNTED)) {
            try {
                File file = new File(Environment.getExternalStorageDirectory(), fileName);
                if (file.exists()) {
                    file.delete();
                }
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
    }

    private static void writeToSDCard(String fileName, String text) {
        if (Environment.getExternalStorageState().equals(Environment.MEDIA_MOUNTED)) {
            FileOutputStream fileOutputStream = null;
            FileInputStream fileInputStream = null;
            try {
                File file = new File(Environment.getExternalStorageDirectory(), fileName);
                fileOutputStream = new FileOutputStream(file, true);
                fileOutputStream.write(text.getBytes());
            } catch (Exception e) {

            } finally {
                if (fileOutputStream != null) {
                    try {
                        fileOutputStream.close();
                    } catch (IOException e) {
                        e.printStackTrace();
                    }
                }
                if (fileInputStream != null) {
                    try {
                        fileInputStream.close();
                    } catch (IOException e) {
                        e.printStackTrace();
                    }
                }
            }
        }
    }

    private static StackTraceElement getTargetStackTraceElement() {
        StackTraceElement targetStackTrace = null;
        boolean shouldTrace = false;
        StackTraceElement[] stackTrace = Thread.currentThread().getStackTrace();
        for (StackTraceElement stackTraceElement : stackTrace) {
            boolean isLogMethod = stackTraceElement.getClassName().equals(Logger.class.getName());
            if (shouldTrace && !isLogMethod) {
                targetStackTrace = stackTraceElement;
                break;
            }
            shouldTrace = isLogMethod;
        }
        return targetStackTrace;
    }

    private static String getJson(String tag, String jsonStr) {
        if (jsonStr == null || jsonStr.isEmpty()) {
            Logger.debug(tag, "Empty/Null json content");
            return "Empty/Null json content";
        }
        try {
            jsonStr = jsonStr.trim();
            if (jsonStr.startsWith("{")) {
                JSONObject jsonObject = new JSONObject(jsonStr);
                Logger.debug(tag, jsonObject.toString(2));
                return jsonObject.toString(2);
            }
            if (jsonStr.startsWith("[")) {
                JSONArray jsonArray = new JSONArray(jsonStr);
                Logger.debug(tag, jsonArray.toString(2));
                return jsonArray.toString(2);
            }
            Logger.debug(tag, "Empty/Null json content");
            return "Invalid Json, Please Check: " + jsonStr;
        } catch (JSONException e) {
            e.printStackTrace();
            Logger.debug(tag, "Empty/Null json content");
            return "Invalid Json, Please Check: " + jsonStr;
        }
    }
}

File: app/src/main/java/com/chichiangho/base/utils/Logger.kt
package com.chichiangho.base.utils

import android.os.Environment
import android.util.Log
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject
import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.io.IOException

object Logger {

    private val ENABLE_LOG_TO_FILE: Boolean = true
    var logLevel = LogLevel.DEBUG // 日志的等级，可以进行配置，最好在Application中进行全局的配置

    enum class LogLevel {
        ERROR {
            override val value: Int get() = 0
        },
        WARN {
            override val value: Int get() = 1
        },
        INFO {
            override val value: Int get() = 2
        },
        DEBUG {
            override val value: Int get() = 3
        };

        abstract val value: Int
    }

    private val MIN_STACK_OFFSET = 3
    private val TOP_LEFT_CORNER = '┌'
    private val BOTTOM_LEFT_CORNER = '└'
    private val MIDDLE_CORNER = '├'
    private val SINGLE_DIVIDER = "────────────────────────────────────────────"
    const val JSON_SPLIT = "||||"
    val TOP_BORDER = TOP_LEFT_CORNER + SINGLE_DIVIDER + SINGLE_DIVIDER
    val BOTTOM_BORDER = BOTTOM_LEFT_CORNER + SINGLE_DIVIDER + SINGLE_DIVIDER
    val MIDDLE_BORDER = MIDDLE_CORNER + SINGLE_DIVIDER + SINGLE_DIVIDER
    val JSON_INDENT = 2

    @JvmStatic fun e(TAG: String, msg: String) {
        if (LogLevel.ERROR.value <= logLevel.value) {
            if (msg.isNotBlank()) {
                val s = getMethodNames()
                Log.e(TAG, String.format(s, msg))
            }
        }
    }

    @JvmStatic fun w(TAG: String, msg: String) {
        if (LogLevel.WARN.value <= logLevel.value) {
            if (msg.isNotBlank()) {
                val s = getMethodNames()
                Log.e(TAG, String.format(s, msg))
            }
        }
    }

    @JvmStatic fun i(TAG: String, msg: String) {
        if (LogLevel.INFO.value <= logLevel.value) {
            if (msg.isNotBlank()) {
                val s = getMethodNames()
                Log.i(TAG, String.format(s, msg))
            }
        }
    }

    @JvmStatic fun d(TAG: String, msg: String) {
        if (LogLevel.DEBUG.value <= logLevel.value) {
            if (msg.isNotBlank()) {
                val s = getMethodNames()
                Log.d(TAG, String.format(s, msg))
            }
        }
    }

    @JvmStatic fun json(TAG: String, json: String) {
        var json = json
        if (json.isBlank()) {
            d(TAG, "Empty/Null json content")
            return
        }

        try {
            val jsonSplit = json.split(JSON_SPLIT).toTypedArray()
            var pre = ""
            if (jsonSplit.size < 2) {
                json = jsonSplit[0]
            } else {
                pre = jsonSplit[0]
                json = jsonSplit[1]
            }

            if (json.startsWith("{")) {
                val jsonObject = JSONObject(json)
                json = (if (pre.isEmpty()) "" else pre + "\n") + jsonObject.toString(JSON_INDENT)
                json = json.replace("\n".toRegex(), "\n| ")
                d(TAG, json)
                return
            }
            if (json.startsWith("[")) {
                val jsonArray = JSONArray(json)
                json = (if (pre.isEmpty()) "" else pre + "\n") + jsonArray.toString(JSON_INDENT)
                json = json.replace("\n".toRegex(), "\n| ")
                d(TAG, json)
                return
            }
            e(TAG, "Invalid Json")
        } catch (e: JSONException) {
            e(TAG, "Invalid Json")
        }
    }

    private fun getMethodNames(): String {
        val sElements = Thread.currentThread().stackTrace
        var stackOffset = getStackOffset(sElements)
        stackOffset++
        val builder = StringBuilder()
        builder.append(TOP_BORDER)
                .append("\r\n") // 添加当前线程名
                .append("| " + "Thread: " + Thread.currentThread().name)
                .append("\r\n")
                .append(MIDDLE_BORDER)
                .append("\r\n") // 添加类名、方法名、行数
                .append("| ")
                .append(sElements[stackOffset].className)
                .append(".")
                .append(sElements[stackOffset].methodName)
                .append(" ")
                .append(" (")
                .append(sElements[stackOffset].fileName)
                .append(":").append(sElements[stackOffset].lineNumber)
                .append(")")
                .append("\r\n")
                .append(MIDDLE_BORDER)
                .append("\r\n") // 添加打印的日志信息
                .append("| ")
                .append("%s")
                .append("\r\n")
                .append(BOTTOM_BORDER)
                .append("\r\n")
        return builder.toString()
    }

    fun getStackOffset(trace: Array<StackTraceElement>): Int {
        var i = MIN_STACK_OFFSET
        while (i < trace.size) {
            val e = trace[i]
            val name = e.className
            if (name != Logger::class.java.name) {
                return --i
            }
            i++
        }
        return -1
    }

    @JvmStatic fun logToFile(time: String, msg: Any?, fileName: String) {
        if (null == msg) {
            return
        }
        if (ENABLE_LOG_TO_FILE) {
            writeToSDCard(fileName, "[" + time + "] " + msg.toString() + "\n")
        }
    }

    private fun writeToSDCard(fileName: String, text: String) {
        if (Environment.getExternalStorageState() == Environment.MEDIA_MOUNTED) {
            var fileOutputStream: FileOutputStream? = null
            val fileInputStream: FileInputStream? = null
            try {
                val file = File(Environment.getExternalStorageDirectory(), fileName)
                fileOutputStream = FileOutputStream(file, true)
                fileOutputStream.write(text.toByteArray())
            } catch (e: Exception) {

            } finally {
                if (fileOutputStream != null) {
                    try {
                        fileOutputStream.close()
                    } catch (e: IOException) {
                        e.printStackTrace()
                    }

                }
                if (fileInputStream != null) {
                    try {
                        fileInputStream.close()
                    } catch (e: IOException) {
                        e.printStackTrace()
                    }

                }
            }
        }
    }
}
